zkvm/src/constraints.rs

```diff
@@ -246,6 +291,16 @@ impl Expression {
             Expression::LinearCombination(terms, _) => r1cs::LinearCombination::from_iter(terms),
         }
     }
+
+    fn to_scalar(&self) -> Option<Scalar> {
```

Comment:
IMHO, we should preserve integerness of the assignment and make it clear that it is the _evaluation_ of the expression, not some sort of alternative representation (`to_*` is usually for conversion methods).
```suggestion
    fn eval(&self) -> Option<ScalarWitness> {
```